src/profile-logic/track-initialization.js

```diff
@@ -0,0 +1,332 @@
+  localTracksByPid: Map<Pid, LocalTrack[]>
+): Map<Pid, TrackIndex[]> {
+  const trackOrderByPid = new Map();
+
+  // Go through each set of tracks, determine the sort order.
+  for (const [pid, tracks] of localTracksByPid) {
+    // Create the default trackOrder.
+    let trackOrder = tracks.map((_, index) => index);
+
+    if (urlTrackOrderByPid !== null) {
+      const urlTrackOrder = urlTrackOrderByPid.get(pid);
+      if (
+        urlTrackOrder !== undefined &&
+        _indexesAreValid(tracks.length, urlTrackOrder)
+      ) {
+        trackOrder = urlTrackOrder;
+      }
+    }
+
+    trackOrderByPid.set(pid, trackOrder);
+  }
+
+  return trackOrderByPid;
+}
+
+export function initializeHiddenLocalTracksByPid(
+  urlHiddenTracksByPid: Map<Pid, Set<TrackIndex>> | null,
+  localTracksByPid: Map<Pid, LocalTrack[]>,
+  threads: Thread[]
+): Map<Pid, Set<TrackIndex>> {
+  const hiddenTracksByPid = new Map();
+
+  // Go through each set of tracks, determine the sort order.
+  for (const [pid, tracks] of localTracksByPid) {
+    const hiddenTracks = new Set();
+
+    if (urlHiddenTracksByPid === null) {
+      for (let trackIndex = 0; trackIndex < tracks.length; trackIndex++) {
+        const track = tracks[trackIndex];
+        if (
+          track.type === 'thread' &&
+          _isThreadIdle(threads[track.threadIndex])
+        ) {
+          hiddenTracks.add(trackIndex);
+        }
+      }
+    } else {
+      const urlHiddenTracks = urlHiddenTracksByPid.get(pid);
+      const trackIndexes = tracks.map((_, index) => index);
+
+      if (urlHiddenTracks !== undefined) {
+        for (const index of urlHiddenTracks) {
+          if (trackIndexes.includes(index)) {
+            hiddenTracks.add(index);
+          }
+        }
+      }
+    }
+
+    hiddenTracksByPid.set(pid, hiddenTracks);
+  }
+
+  return hiddenTracksByPid;
+}
+
+export function computeTracksByPid(profile: Profile): Map<Pid, LocalTrack[]> {
+  const localTracksByPid = new Map();
+
+  for (
+    let threadIndex = 0;
+    threadIndex < profile.threads.length;
+    threadIndex++
+  ) {
+    const thread = profile.threads[threadIndex];
+    const { pid } = thread;
+    // Get or create the tracks and trackOrder.
+    let tracks = localTracksByPid.get(pid);
+    if (tracks === undefined) {
+      tracks = [];
+      localTracksByPid.set(pid, tracks);
+    }
+
+    if (_isMainThread(thread)) {
+      // This thread was already added as a GlobalTrack.
+      continue;
+    }
+
+    tracks.push({ type: 'thread', threadIndex });
+  }
+
+  return localTracksByPid;
+}
+
+export function computeGlobalTracks(profile: Profile): GlobalTrack[] {
+  // Defining this ProcessTrack type here helps flow understand the intent of
+  // the internals of this function.
+  type ProcessTrack = {
+    type: 'process',
+    pid: Pid,
+    mainThreadIndex: number | null,
+  };
+  const globalTracksByPid: Map<Pid, ProcessTrack> = new Map();
+  const globalTracks: GlobalTrack[] = [];
```

Comment:
I think you can do without this variable, and `return [...globalTracksByPid.values()]` at the end. This works because `Map.prototype.values` iterates in the insertion order. This should simplify a tiny bit this function.